### ISSUE TITLE
[FEATURE] Upgrade: retry image downloading

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#3144]
**Is your feature request related to a problem? Please describe.**

We can improve the upgrade start experience if there is an upgrade image downloading failure.
We have seen issues like:
- https://github.com/harvester/harvester/issues/2104
- And user also reports failure:
  ```
  Status:
    Applied URL:  https://releases.rancher.com/harvester/v1.1.1-rc1/harvester-v1.1.1-rc1-amd64.iso
    Conditions:
      Last Update Time:  2022-11-10T20:37:09Z
      Message:           Head "https://releases.rancher.com/harvester/v1.1.1-rc1/harvester-v1.1.1-rc1-amd64.iso": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
      Status:            False
      Type:              Initialized
    Storage Class Name:  longhorn-harvester-iso-5wxpd
  Events:                <none>
  ```

**Describe the solution you'd like**

- We implement our retry logic in the upgrade controller.
- Longhorn support backing image retry.

**Describe alternatives you've considered**

**Additional context**

**Comment by TachunLin** [comment]
#### [PASS] Case 1: Don't retry unlimitedly for a failed URL.

```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineImage
metadata:
  generateName: image-
  namespace: default
spec:
  checksum: ""
  displayName: focal-1.img
  pvcName: ""
  pvcNamespace: ""
  retry: 1
  sourceType: download
  storageClassParameters:
    migratable: "true"
    numberOfReplicas: "3"
    staleReplicaTimeout: "30"
  url: http://192.168.122.181:8000/focal-server-cloudimg-amd64.img
```

```
n1-112rc1:~ # kubectl get vmimage -A -o yaml -w
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineImage
...
status:
  appliedUrl: http://192.168.122.181:8000/focal-server-cloudimg-amd64.img
  conditions:
  - lastUpdateTime: "2023-02-21T10:22:24Z"
    message: VMImage has reached the specified retry limit
    status: "True"
    type: RetryLimitExceeded
  - lastUpdateTime: "2023-02-21T10:22:24Z"
    message: 'Head "http://192.168.122.181:8000/focal-server-cloudimg-amd64.img":
      dial tcp 192.168.122.181:8000: connect: no route to host'
    status: "False"
    type: Initialized
  failed: 2
  lastFailedTime: "2023-02-21T10:22:24Z"
```
```
n1-112rc1:~ # kubectl delete vmimage image-9h4gb
virtualmachineimage.harvesterhci.io "image-9h4gb" deleted
```

#### [PASS] Case 2: Retry succeed.

![image](https://user-images.githubusercontent.com/29251855/220324291-3ca197a5-8ee3-48a2-bac7-1b0247f09eed.png)

![image](https://user-images.githubusercontent.com/29251855/220325009-c0269db9-d687-4359-bb90-a556c81a83e5.png)


```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineImage
metadata:
  generateName: image-
  namespace: default
spec:
  checksum: ""
  displayName: focal-3.img
  pvcName: ""
  pvcNamespace: ""
  retry: 3
  sourceType: download
  storageClassParameters:
    migratable: "true"
    numberOfReplicas: "3"
    staleReplicaTimeout: "30"
  url: http://192.168.122.1:8000/focal-server-cloudimg-amd64.img
```

```
---
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineImage
metadata:
  creationTimestamp: "2023-02-21T10:45:53Z"
  finalizers:
  - wrangler.cattle.io/vm-image-controller
  generateName: image-
  generation: 5
  labels:
    harvesterhci.io/imageDisplayName: focal-3.img
  name: image-rwm2b
  namespace: default
  resourceVersion: "90753"
  uid: 45f30aa6-6c8b-43de-9cb5-4aebaa19e21b
spec:
  checksum: ""
  displayName: focal-3.img
  pvcName: ""
  pvcNamespace: ""
  retry: 3
  sourceType: download
  storageClassParameters:
    migratable: "true"
    numberOfReplicas: "3"
    staleReplicaTimeout: "30"
  url: http://192.168.122.1:8000/focal-server-cloudimg-amd64.img
status:
  appliedUrl: http://192.168.122.1:8000/focal-server-cloudimg-amd64.img
  conditions:
  - lastUpdateTime: "2023-02-21T10:46:27Z"
    reason: Imported
    status: "True"
    type: Imported
  - lastUpdateTime: "2023-02-21T10:46:13Z"
    reason: Initialized
    status: "True"
    type: Initialized
  failed: 2
  lastFailedTime: "2023-02-21T10:46:03Z"
  progress: 100
  size: 579665920
  storageClassName: longhorn-image-rwm2b

```


#### [PASS] Case 3: VMimage works after upgrade from v1.1.0 
* Can upgrade from v1.1.0 to v1.1.2-rc1 correctly
  ![image](https://user-images.githubusercontent.com/29251855/220609059-6b25edb5-59b2-469f-abba-0a885767c17f.png)

* The original VMImage work as expected
  ![image](https://user-images.githubusercontent.com/29251855/220609602-7d008d69-5993-43d9-bd54-1f832ae6a048.png)

* Create a new VMImage and attach to VM can work as expected
  ![image](https://user-images.githubusercontent.com/29251855/220611046-e2dc95a6-50d8-4736-9cc2-befb118bd99b.png)
  ![image](https://user-images.githubusercontent.com/29251855/220611407-9979c41c-902e-4c4e-b828-517361b05805.png)


#### [PASS] Case 4: Users can't update URL for existent images 

* Update the URL of existing downloaded image, while saving webhook should prompt an error. `spec.url: url cannot be modified` 
  ```
  node1:~ # kubectl get vmimages
  NAME          DISPLAY-NAME                                              SIZE         AGE
  image-5z8tz   centos-8-genericcloud-8.4.2105-20210603.0.x86_64.qcow2    1309652992   23m
  image-6h5gh   focal-server-cloudimg-amd64.img                           645922816    133m
  image-f9rt4   opensuse-leap-15.3.x86_64-1.0.1-nocloud-build2.97.qcow2   568066048    131m
  node1:~ # kubectl edit vmimages image-6h5gh
  ```
  
  ```
  spec:
    checksum: ""
    displayName: focal-server-cloudimg-amd64.img
    pvcName: ""
    pvcNamespace: ""
    retry: 3
    sourceType: download
    storageClassParameters:
      migratable: "true"
      numberOfReplicas: "3"
      staleReplicaTimeout: "30"
    url: https://cloud-images.ubuntu.com/focal/update/focal-server-cloudimg-amd64.img
  
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # virtualmachineimages.harvesterhci.io "image-6h5gh" was not valid:
  # * spec.url: url cannot be modified
  #
  
  ```

#### [PASS] Case 5: Upgrade controller can show RetryLimitExceeded error
Upgrade on `v1.1.2-rc1` with non-existent ISO URL. The dashboard can prompt error message
  ```
  VMImage has reached the specified retry limit
  ```


  ![image](https://user-images.githubusercontent.com/29251855/220638101-0d7acd72-3929-473a-94ad-11b4337ec7d3.png)
  
  ![image](https://user-images.githubusercontent.com/29251855/220639000-92f2fe46-2e07-4554-b75e-5d7c2c388ed9.png)

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.1.2-rc1`. Close this issue.

### Result
Verified can pass all test scenarios of retry image download in upgrade 
https://github.com/harvester/harvester/issues/3144#issuecomment-1438231006

- [x]  Case 1: Don't retry unlimitedly for a failed URL
- [x]  Case 2: Retry succeed.
- [x]  Case 3: VMimage works after upgrade from v1.1.0
- [x]  Case 4: Users can't update URL for existent images
- [x]  Case 5: Upgrade controller can show RetryLimitExceeded error


### Test Information
* Test Environment: 3 nodes harvester on local bare metal macines
* Harvester version: v1.1.0 to v1.1.2-rc1

### Verify Steps
https://github.com/harvester/tests/issues/610#issue-1449309815
